esp/platform/txsummary.hpp

```diff
@@ -50,80 +55,300 @@ class CTxSummary : extends CInterface
+    // named 'sourceKey'
+    virtual bool setCopyValueOf(const char* destKey, const char* sourceKey, const LogLevel logLevel = LogMin, const unsigned int group = DEFAULT_GROUP);
 
-    void log(const LogLevel logLevel);
+    void log(ITxSummaryProfile* profile, const LogLevel logLevel, const unsigned int group = DEFAULT_GROUP, const unsigned int requestedStyle = TXSUMMARY_OUT_TEXT);
 
     // Fetches an existing or new instance of a named CumulativeTime. The name
-    // must not be NULL or empty, but it may be a duplicate of an existing
-    // key-value pair. Duplication, while permitted, should be avoided due to
-    // the potential for confusion.
-    virtual CumulativeTimer* queryTimer(const char* name);
+    // must not be NULL or empty, duplication is not permitted
+    virtual CumulativeTimer* queryTimer(const char* key);
 
     // Adds the given milliseconds to an existing or new named CumulativeTimer.
     // The same conditions as for getTimer apply.
-    virtual bool updateTimer(const char* name, unsigned long long delta, const LogLevel logLevel = LogMin);
+    virtual bool updateTimer(const char* name, unsigned long long delta, const LogLevel logLevel = LogMin, const unsigned int group = DEFAULT_GROUP);
 
 protected:
     // Log the summary contents on destruction.
     ~CTxSummary();
 
 private:
+    struct TxEntryBase;
+
+    virtual void set(const char* key, TxEntryBase* value);
+
+    // Returns true if the summary contains an entry matching key
     bool __contains(const char* key) const;
 
-    struct Entry
+    // Returns the entry matching key if found, or nullptr otherwise
+    TxEntryBase* queryEntry(const char* key);
+
+    // Each entry is a subclass of TxEntryBase that allows us
+    // to keep type information along with the value to use for
+    // serialization. Each Entry has these primary attributes:
+    //
+    //  name
+    //      The name when serialized. It may be in the form
+    //      of a simplified dot-delimited JSON-style path.
+    //      That form is used to give structure when
+    //      serializing to JSON. The implementation tracks
+    //      the name in two ways- as 'name' and 'fullname'
+    //      to aid with serialization.
+    //
+    //  logLevel
+    //      The minimum logLevel that must be requested during
+    //      serialization in order for this element to appear
+    //      in the output.
+    //
+    //  group
+    //      A bit flag indicating which groups this element
+    //      belongs to. Serialization can specify which groups
+    //      are included in output
+    //
+    //  value
+    //      Subclasses of TxEntryBase include a value member of
+    //      a type appropriate for the entry.
+    //
+    struct TxEntryBase : public CInterface
     {
-        StringAttr key;
-        StringAttr value;
+        // The complete name of the entry, may be a dot-delimited
+        // path.
+        StringAttr fullname;
```

Comment:
Why StringAttr instead of StringBuffer? Do we really want NULL keys?

Reply:
I think I thought that StringAttr was a preferred variant here- the key and value were StringAttr previously.

Reply:
In light of changing the name to `const char*`, I'm keeping then StringAttr, but using the `str()` accessor.